### DISCUSSION TITLE
Maho SQL upgrade scripts

### DISCUSSION THREAD

**Discussion by justinbeaty** [MahoCommerce/maho#37] [Ideas]
> about the install scripts, do you think they fit the mage_core or mage_customer module better? we will have to bump the version number of a module and that eventually will cause conflicts with openmage. Obviously we can manage those conflicts manually when they will happen but, do you have any idea if this situation would be avoidable somehow? https://github.com/MahoCommerce/maho/pull/36#issuecomment-2387956958

How about adding support for `app/code/*/Module_Name/sql/module_setup/maho-install-xx.y.z` files? At first I thought it wasn't a good idea, but then thought about it some more:

1. Add new column to `core_resource` called `data_version_maho`.
2. In `Mage_Core_Model_Resource_Setup`, load these files if `data_version_maho < xx.y.z <= Mage::app()->getMahoVersion()`
3. Update `data_version_maho` to the highest `xx.y.z` that ran for that particular resource. Most rows in that table will have `data_version_maho == null` for sometime unless that resource has a maho-install script.

Some thoughts:

- The current `getMahoVersion()` is 24.11.0, so if I wanted  to add an install script in a PR I might name my file `app/code/core/Mage_Customer/sql/customer_setup/maho-install-24.10.0.php` since it's still October. By using the current month instead of the next expected version, you have flexibility in the release schedule. I.e. if you had named it `maho-install-24.11.0.php`, but then needed to make a release in October, the script wouldn't run
- For multiple PRs to the same resource setup, you could either use `24.10.1`, or just merge the two scripts when merging the main branch into the PR. The former would be a problem if it's the release month, since `24.11.1` wouldn't be run for the `24.11.0` release. So probably stick with the merging them. 
- We don't need to use a from and to version in the maho-install files. I can't think of any case where we have ever wanted to use a from version lower than the current release. It can cause a discrepancy anyway depending on if the user upgrades with every version vs sporadically.
- We don't have to keep bumping those arbitrary `<version>` nodes in the module's `config.xml`

Edit: clarified that each module can have their own maho-install scripts, instead of just one huge place for them.

**Comment by justinbeaty** [comment]
If you've ever seen [this video](https://www.youtube.com/watch?v=AbSehcT19u0), it perfectly sums up the experience right now.

Working on the EAV code, but first need to update the persistent module, but then need to figure out install script, but then...

**Comment by fballiano** [comment]
I think it's a great idea as usual, I started coding it, I hope I'll be able to show you something super soon!

**Reply by justinbeaty** [reply]
> I think it's a great idea as usual, I started coding it, I hope I'll be able to show you something super soon!

Sweet!

Also I never knew in all these years, but Magento has two locations for upgrade scripts: `Module_Name/sql` and `Module_Name/data`. The first one being for creating and altering tables. The second one being for adding or modifying data, either with `core_write`, or with Magento models i.e. `Mage::getModel('catalog/category')->load(1)->setName('Root Catalog')->save()`.

You very well may have already known that, but anytime I've written an internal module for my store I've just put it all in the `sql` directory. In any case, maybe we need two columns, `version_maho` and `data_version_maho`.

**Reply by fballiano** [reply]
yep I knew about the "data" directory but actually I don't know, it was added in magento because the necessities are bigger, but in our case I think only 1 folder is enough at the moment no?

**Reply by justinbeaty** [reply]
One folder is enough for us, for sure. I only thought about it in case someone put a `maho-upgrade-xx.y.z` script in `data` and didn't know why it wasn't working.

**Reply by justinbeaty** [reply]
Also, probably only core modules need the maho scripts. Normal module developers can use the old way.

**Reply by fballiano** [reply]
if we do maho-24.9.0.php, maho-24.9.1.php etc and they are run in sequence after the magento ones... I think we don't need much more, also because I hope we won't be changing the database that much :-D

**Reply by justinbeaty** [reply]
I just realized that if you want to have just one set of upgrade scripts for all modules, can't we just make a `app/code/core/Maho/Core` module? Pin the ordering in `app/code/core/Mage/Core/Model/Config.php` like was done with other core modules.

A more custom solution is good too, though. **Edit:** actually, a custom solution is probably better so we don't have to stuff everything into one `upgrade-24.9.0-24.11.0.php` script... Carry on. 😄

**Reply by fballiano** [reply]
I thought about it and it would be easier but I don't know... we'd need to load it after the rest of the modules (which is kinda weird) and it's "another module" when we're trying to remove them... and we could add a column to a table part of a disabled module... so all and all I think it's cleaner to have a "setup script" dedicated to maho for every module.

makes sense?

**Reply by justinbeaty** [reply]
Yup, makes sense and forget I mentioned it!

**Reply by fballiano** [reply]
I've created https://github.com/MahoCommerce/maho/pull/38

**Comment by colinmollenhour** [comment]
If you're willing to make a major change, I suggest considering moving away from version numbers and instead use dates or arbitrary strings. There are cases where dependencies between modules cause problems such that updates belonging to modules A and B need to be applied in a specific order but using only dependencies and version numbers all of A has to be applied before any of B. Using dates you can say that A.1 applies, then B.1, then A.2, etc. Most if not all mature migration libraries use dates or arbitrary sortable strings like this for a long time. You typically insert a row for each migration rather than just store the latest one so that if a feature with an earlier sort order is merged later, you don't have to rename any files. This has a lot of advantages on busy projects and if a convention is used consistently like a timestamp format it makes it possible to solve dependency issues naturally.

You could maintain BC for existing migrations to the new system without having to rename them, keeping compat with third-party modules. For example, say that all new migrations under the new system could be named like `2024-10-3-my-upgrade.php` and it would be assumed that all `upgrade-X-Y.php` files come first.

- Apply all version numbers for each module, one module at a time using dependency tree sorting
- Then apply all non-numbered upgrades in global sort order

**Reply by fballiano** [reply]
it is an interesting idea (it would be more similar to laravel migrations, which I like since it's kinda the default standard nowadays) but I don't understand how to keep compatibility with 3rd party modules.

@justinbeaty what do you think about this?

**Reply by justinbeaty** [reply]
If I'm understanding this correctly, the only two changes to make are:

1) Create a new table `maho_setup` with at least two columns, `code` and `script_id`. Whenever a migration script is ran, insert a row with it's resource name, and the arbitrary string.

2) In `applyAllMahoUpdates()`, find all `app/code/*/*/*/sql/maho_setup/*.php` files, filter out those already in the table, sort them, and then run them (making sure we still use the correct setup class).

---

> Apply all version numbers for each module, one module at a time using dependency tree sorting

I had to double check that Mage already runs migration files based on module dependency, and it seems it does, see this test: https://github.com/justinbeaty/maho/commit/52aa197ebb4dd60b9dd003d15f9fc7e99bd742dc.

Since Test_A depends on Test_B, it outputs:

```
2024-10-03T19:46:45+00:00 DEBUG (7): test_b_resource
2024-10-03T19:46:45+00:00 DEBUG (7): test_a_resource
```

But remove the dependency, and it outputs:

```
2024-10-03T19:47:08+00:00 DEBUG (7): test_a_resource
2024-10-03T19:47:08+00:00 DEBUG (7): test_b_resource
```

So regarding compatibility with 3rd party modules and changes to setup scripts in OpenMage, I don't think there's anything we need to do there.

**Reply by justinbeaty** [reply]
One concern I have is it would run Maho setup scripts in different orders for different users, depending on how often they upgrade. Such as in the example given by Colin, if feature B is merged in 24.9.0 and feature A is merged in 24.11.0, then those who upgraded to 24.9.0 run B before A. But if you skipped 24.9.0, you run A before B.

Theoretically, it shouldn't matter because if one script depended on another, they should have appropriate names. I assume this was considered for larger projects using this method, and deemed acceptable.

**Reply by colinmollenhour** [reply]
That happens already. If you have A with three versions and B with three versions and B depends on A, you'll get A1, A2, A3, B1, B2, B3. Now if you add A4, A4 will run after B3. But on a *new* installation, A4 will run *before* B1. At least with the date-naming it is fairly likely that on a new installation A4 would run last, but you are right that it is not guaranteed unless you had a policy to update any "old" timestamps just before merging which is quite feasible.

**Reply by justinbeaty** [reply]
You are right, so then the timestamp method is definitely less bad than the current behavior.

Was I missing anything about the backwards compatibility issue? The PR @fballiano made already does run the maho scripts after all "legacy" scripts have run, so it's really just about ordering the maho ones better.

**Reply by fballiano** [reply]
I think that the timestamp based solution would be very interesting but
- it would require a huge work of converting/testing all the old install/data scripts (a lot of time we don't have with low returned value)
- it's true that those files are never updated but it would great a great deal of incompatibility between the codebases that, at the moment, I would avoid

@justinbeaty this would mean leaving everything as it already is now, ironing out the couple of bungs you find right?

**Reply by fballiano** [reply]
BTW thanks everybody for the great ideas

**Reply by justinbeaty** [reply]
We can leave the old install scripts as they are, so that it doesn't impact the way openmage or 3rd party modules work.

It would just be changing the way maho scripts are loaded. Currently the execution order is:

1. `Mage/Core/sql/maho_setup/maho-24.10.0.php`
2. `Mage/Core/sql/maho_setup/maho-24.10.1.php`
3.  `Mage/Checkout/sql/maho_setup/maho-24.10.0.php`

Colin proposes it should be:
1. `Mage/Core/sql/maho_setup/maho-24.10.0.php`
2.  `Mage/Checkout/sql/maho_setup/maho-24.10.0.php`
3. `Mage/Core/sql/maho_setup/maho-24.10.1.php`

---

I think we have three options:

1. Leave as it is now, just fix bugs.

2. Change the load order, but don't create a table to track every ran script. Just keep recording the highest `maho-xx.y.z.php` that ran in `core_resource` like your implementation already does. PRs must rename maho upgrade scripts to account for earlier merged scripts. I.e. don't merge `maho-24.10.1.php` before  `maho-24.10.0.php`.

3. Change the load order, and track every ran script in a new table. This would allow `maho-24.10.1.php` to be run first, and still later run `maho-24.10.0.php`. Load order would be different if the user upgrades vs fresh install.

I think that since Maho versions are already dates, we can implement any of these at a later time and stay BC, *as long as* we don't do stuff like make `Mage/Checkout/sql/maho_setup/24.10.0` rely on changes in `Mage/Core/sql/maho_setup/24.10.1` in the meantime.

I would probably vote for option 1 right now, and option 2 at a later time.

**Reply by fballiano** [reply]
I think that, for the foreseeable future, option 1 is ok, we should refrain from modifying the db too much anyway and having that freedom could even be worse 😅
